### PR TITLE
step selector: gather data for data table

### PR DESCRIPTION
**Pull request by JamesHollyer**
* Motivation for features / changes
This is one pr in a chain intended to add a data table feature. This change adds data types and gathers the data needed for the table. For more context on where this is headed see https://github.com/tensorflow/tensorboard/pull/5743

* Technical description of changes
The ScalarCard now decides which columns we will have on the table(currently this is hard coded but, we plan to allow for the user to select which columns they want so this is built with that in mind). It also uses the selectedTime to decide which points of data will be displayed.  Eventually we will replace selected time with a more narrow step selection that is specific for that card. 

* Screenshots of UI changes
This PR makes no UI changes but, for reference we are aiming to make a table like this:
Screen Shot 2022-05-23 at 4 29 28 PM

* Detailed steps to verify changes work correctly (as executed by you)

* Alternate designs / implementations considered
There was another design to pass the point data and run metadata down to the table and have that table understand how to get the correct data per column. We plan to one day allow for ranges of data and there are some calculations that would prove very difficult in the data table(for instance the average of all values within the range) so we went with this solution where the values are "calculated" in the ScalarCardComponent and then passed to the data table.

### CODE REVIEW

File: tensorboard/webapp/widgets/data_table/data_table_component.ts [hunk 1]
```diff
@@ -13,12 +13,19 @@ See the License for the specific language governing permissions and
...
 
 @Component({
   selector: 'tb-data-table',
   templateUrl: 'data_table_component.ng.html',
   styleUrls: ['data_table_component.css'],
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class DataTableComponent {}
+export class DataTableComponent {
+  @Input() headers!: ColumnHeaders[];
```
Comment (japie1235813) [inline tensorboard/webapp/widgets/data_table/data_table_component.ts:29]: The order of `ColumnHeaders[]` matters on component rendering but it's not documented anywhere, can we make a note here?

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_component.ts [hunk 2]
```diff
@@ -97,6 +100,12 @@ export class ScalarCardComponent<Downloader> {
 
   yScaleType = ScaleType.LINEAR;
   isViewBoxOverridden: boolean = false;
+  dataHeaders: ColumnHeaders[] = [
+    ColumnHeaders.RUN,
+    ColumnHeaders.VALUE,
+    ColumnHeaders.STEP,
+    ColumnHeaders.RELATIVE_TIME,
+  ];
```
Comment (japie1235813) [inline tensorboard/webapp/metrics/views/card_renderer/scalar_card_component.ts:108]: Just a thought, do you expect this array to be dynamic changing or a constant? 

Not required in this PR but I might prefer is as as a constant and put this outside of the scalar_card_component. Later in range selection, we put more attributes in the array but the data_table_component is smart enough to not render the column if there is no such key provided in data (type SelectedStepRunData). That is to say, in single selection  we have 'VALUE_CHANGE' in 'headers' but not in keys of SelectedStepRunData type object so the data_component will not render this column.

**Review by japie1235813 — APPROVED** [review]
woo-hoo I like this structure! Great work. Just some nits.

File: tensorboard/webapp/metrics/views/card_renderer/scalar_card_types.ts [hunk 3]
```diff
@@ -70,3 +84,7 @@ export interface PartitionedSeries {
   runId: string;
   points: ScalarCardPoint[];
 }
+
+export type SelectedStepRunData = {
```
[the remark this answers is not in the document]
Comment (JamesHollyer) [reply tensorboard/webapp/metrics/views/card_renderer/scalar_card_types.ts:93]: Done.

File: tensorboard/webapp/widgets/data_table/data_table_component.ts (hunk 1 above)
Comment (JamesHollyer) [reply tensorboard/webapp/widgets/data_table/data_table_component.ts:31]: Done.

**Review by JamesHollyer — COMMENTED** [review]
Thanks Jie-Wei

**Pull request merged**